update dispatcher url matcher docs (#228)
front-commerce/developers.front-commerce.com: source/docs/advanced/theme/route-dispatcher.md

**Magento:** If the url you are trying to add are managed by Magento, you don't need any of this. You should instead [add an url rewrite directly in your backend](https://devdocs.magento.com/guides/v2.2/cloud/configure/import-url-rewrites.html), since the mechanism already exists in the Magento module of Front-Commerce.
</blockquote>

To add a new `urlMatcher` to a module you need to register `FrontCommerce/Core` as a module dependecy and then use `registerUrlMatcher` to add a custom `urlMatcher`:

```diff
import typeDefs from "./schema.gql";
+import myCustomUrlMatcher from './urlMatcher.js';
import MyCustomLoader from './loader.js';
import { makeUserClientFromRequest } from "server/modules/magento2/core/factories";

const moduleDefinition = {
  namespace: "MyCustomModule",
-  dependencies: ["Front-Commerce/Core"],
+  dependencies: [],
  typeDefs,
  contextEnhancer: ({ req, loaders }) => {
    const axiosInstance = makeUserClientFromRequest(req);
    const CustomLoader = MyCustomLoader(axiosInstance);
+    loaders.Page.registerUrlMatcher(myCustomUrlMatcher(CustomLoader));
    return {
      MyCustomLoader: CustomLoader
    };
  },
};

export default moduleDefinition;
```

> ```diff

Reviewer: For whole files, I can't see any benefits to `diff` syntax.
Here, `js` might lead to more readable example IMO. I suggest to rework the example.

// urlMatcher.js
+const myCustomUrlMatcher = (loader) => async (url, match) => {

> +  if(match.__typename) {

Reviewer: CS in this example needs to be reworked.
We must try to keep it as standard as possible.

I suggest at least:
- pass it into Prettier (either in your editor or in https://prettier.io/playground/)
- ensure the code is _valid_ (e.g: `<canonical url>` should instead use a property of `myMatch`)
- prevent looooooong lines: either remove comments (over a paragraph below) or put them on another line(s)… refactoring into temporary variables can also make things more expressive and remove the need for some comments

+    // a match have been already loaded (you can still override it or return it as is like we do here)
+    return match;
+  }
+  const myMatch = await loader.loadByUrl(url);
+  if(!myMatch) {
+    return match; // return original match if none is found (this is needed!)
+  }
+  return { // return your custom routable entity
+    path: <canonical_url>,      // depending on your use case can be same as url
+    url,
+    __typename: 'MyCustomType', // the GraphQL type of your custom type (should be added in schema.gql),
+    ...myMatch, // or some adaptation of it. this will be available to your component defined bellow on a prop called `matched`
+  }
+};
+
+export default myCustomUrlMatcher;
```

```diff
# schema.gql

> +MyCustomType {

Reviewer: This type must implements the `Routable` interface and thus have à `path: String` field.

I think it should be part of the example.

+  title: String! # example field
+  ... # define own schema
+}
```

```diff
// loader.js
const MyCustomLoader = (axiosInstance) => {
  return {
+    loadByUrl: (url) => {

> +      return ( // example: get data by issuing an external api call

Reviewer: Same feedbacks as above about CS in this code (indentation and _executable_ call). It could also be a `js` block instead of a `diff`.

Here for instance, the comment could be removed if an expressive URL is used in `axiosInstance.get` instead of `...` (suggestion: `/my/api/my-custom-entity/${url}`).

+        axiosInstance.get(...)
+        .then((response) => response.data)
+        .catch((err) => err.response?.status === 404 ? null : Promise.reject(err));
+      );
+    },
  };
};
```

>     <blockquote class="note">
>     See [Loaders](/docs/advanced/graphql/slim-down-resolvers-with-loaders.html) documentation to learn how to instantiate your loader and add it to your GraphQL context.
>     </blockquote>

Reviewer: ```suggestion
<blockquote class="note">
    See [Loaders](/docs/advanced/graphql/slim-down-resolvers-with-loaders.html) documentation to learn how to instantiate your loader and add it to your GraphQL context.
</blockquote>
```

Indentation led to an automatic code block instead:
![Selection_001](https://user-images.githubusercontent.com/75968/128192251-5de9cfe2-103e-469f-be17-0a89bffa8f74.jpg)

> Once you've done, you should be able to test that everything works as expected by using the GraphQL Playground at http://localhost:4000/playground and executing the following query:

Reviewer: ```suggestion
Once this is done, you should be able to test that everything works as expected by using the GraphQL Playground at http://localhost:4000/playground and executing the following query:
```

```gql
{
  route(url: "some-dynamic-custom-url") {
    url

>     title # example

Reviewer: `path` is the only field from the `Routable` interface.

This example should thus be reworked with something close to:

```graphql
{
  route(url: "some-dynamic-custom-url") {
    path
	... on MyCustomType {
	    title # from the example schema above
 	}
  }
}
```

>     ...   # custom type fields

Reviewer: ```suggestion
```
not needed as the `title` should be sufficient. It makes the example executable without syntax error.

In Front-Commerce’s core integrations (such as Magento2), the association between a URL and a page is already implemented for entities like Products, Categories, CMS pages… But depending on your own site, you might need to add new ones.

To do so, you will need to proceed in two steps:

- Register a new urlMatcher

> - Add the mapping between the type returned by `route` and the page component that should be displayed

Reviewer: ```suggestion
- Add GraphQL type to the dispatcher query
- Add the mapping between the type returned by `route` and the page component that should be displayed
```

It'd be perfect with anchor links to the related section too.

## Register a new urlMatcher

<blockquote class="info">
**Magento:** If the url you are trying to add are managed by Magento, you don't need any of this. You should instead [add an url rewrite directly in your backend](https://devdocs.magento.com/guides/v2.2/cloud/configure/import-url-rewrites.html), since the mechanism already exists in the Magento module of Front-Commerce.
</blockquote>

To add a new `urlMatcher` to a module you need to register `FrontCommerce/Core` as a module dependecy and then use `registerUrlMatcher` to add a custom `urlMatcher`:

```diff
import typeDefs from "./schema.gql";
+import myCustomUrlMatcher from './urlMatcher.js';
import MyCustomLoader from './loader.js';
import { makeUserClientFromRequest } from "server/modules/magento2/core/factories";

const moduleDefinition = {
  namespace: "MyCustomModule",

> -  dependencies: ["Front-Commerce/Core"],
> +  dependencies: [],

Reviewer: It should be the opposite:
```suggestion
-  dependencies: [],
+  dependencies: ["Front-Commerce/Core"],
```

    path
    ... on MyCustomType {
      title # example
    }
  }
}
```

## Add GraphQL type to the dispatcher query

Override `DispatcherQuery` if you have not already done so. You can find it under `src/web/theme/modules/Router/DispatcherQuery.gql`.

Add an [inline fragment](https://graphql.org/learn/queries/#inline-fragments) to the `DispatcherQuery` for your newly created GraphQL type:

```diff
...
+import 'path_to_your_fragment/MyCustomTypeFragment.gql'
+import 'path_to_your_fragment/MyCustomTypeCacheControlFragment.gql'
query MatchUrls($url: String!, $params: QueryInput) {
  route(url: $url) {
    path
    __typename
    ... on RedirectEntity {
      redirectTo
      redirectType
    }
    ... on Product {
      ...ProductFragment
      ...ProductCacheControlFragment
    }
    ... on Category {
      ...CategoryFragment
      ...CategoryCacheControlFragment
    }
    ... on CmsPage {
      ...CmsPageFragment
      ...CmsPageCacheControlFragment
    }
+    ... on MyCustomType {
+      ...MyCustomTypeFragment
+      ...MyCustomTypeCacheControlFragment # optional check note below for link
+    }
  }
}
```

```gql
# MyCustomTypeFragment.gql

> fragment AlbumFragment on Album {

Reviewer: ```suggestion
fragment MyCustomTypeFragment on MyCustomType {
```
